src/database.c

```diff
@@ -419,25 +449,34 @@ _add_to_db(ProfMessage* message, char* type, const Jid* const from_jid, const Ji
 
         sqlite3_stmt* lmc_stmt = NULL;
 
-        if (SQLITE_OK == sqlite3_prepare_v2(g_chatlog_database, replace_check_query, -1, &lmc_stmt, NULL)) {
-            if (sqlite3_step(lmc_stmt) == SQLITE_ROW) {
-                const char* from_jid_orig = (const char*)sqlite3_column_text(lmc_stmt, 0);
+        if (SQLITE_OK != sqlite3_prepare_v2(g_chatlog_database, replace_check_query, -1, &lmc_stmt, NULL)) {
+            log_error("SQLite error in _add_to_db() on selecting original message: %s", sqlite3_errmsg(g_chatlog_database));
+            return;
+        }
+
+        if (sqlite3_step(lmc_stmt) == SQLITE_ROW) {
+            original_message_id = sqlite3_column_int(lmc_stmt, 0);
+            const char* from_jid_orig = (const char*)sqlite3_column_text(lmc_stmt, 1);
+
+            // Handle non-XEP-compliant replacement messages (edit->edit->original)
+            int tmp = sqlite3_column_int(lmc_stmt, 2);
+            original_message_id = tmp ? tmp : original_message_id;
 
-                if (g_strcmp0(from_jid_orig, from_jid->barejid) != 0) {
-                    log_error("Mismatch in sender JIDs when trying to do LMC. Corrected message sender: %s. Original message sender: %s. Replace-ID: %s. Message: %s", from_jid->barejid, from_jid_orig, message->replace_id, message->plain);
-                    cons_show_error("%s sent message correction with mismatched sender. See log for details.", from_jid->barejid);
-                    sqlite3_finalize(lmc_stmt);
-                    return;
-                }
+            if (g_strcmp0(from_jid_orig, from_jid->barejid) != 0) {
+                log_error("Mismatch in sender JIDs when trying to do LMC. Corrected message sender: %s. Original message sender: %s. Replace-ID: %s. Message: %s", from_jid->barejid, from_jid_orig, message->replace_id, message->plain);
+                cons_show_error("%s sent a message correction with mismatched sender. See log for details.", from_jid->barejid);
+                sqlite3_finalize(lmc_stmt);
+                return;
             }
-            sqlite3_finalize(lmc_stmt);
+        } else {
+            log_warning("Got LMC message that does not have original message counterpart in the database from %s", message->from_jid->fulljid);
         }
+        sqlite3_finalize(lmc_stmt);
     }
 
     // Check for duplicate messages
-    auto_sqlite char* duplicate_check_query = sqlite3_mprintf("SELECT 1 FROM `ChatLogs` WHERE (`archive_id` = '%q' AND `archive_id` != '') OR (`stanza_id` = '%q' AND `stanza_id` != '')",
-                                                              message->stanzaid ? message->stanzaid : "",
-                                                              message->id ? message->id : "");
+    auto_sqlite char* duplicate_check_query = sqlite3_mprintf("SELECT 1 FROM `ChatLogs` WHERE (`archive_id` = '%q' AND `archive_id` != '')",
+                                                              message->stanzaid ? message->stanzaid : "");
```

Comment:
Shouldn't we maybe handle the case where `message->stanzaid == ("" or NULL)` before getting here? FMU this should never happen!

Then the `AND 'archive_id' != ''` part could be dropped.